Race condition between copying Dockerfile and tarring context

I discovered a race condition today between copying the image Dockerfile to the project root, and tarring of the context.

Basically if you have a _really_ small context, the tar action completes before the Dockerfile has had chance to copy to the project root, causing the following error,

```
Error: HTTP code is 500 which indicates error: server error - null
    at Modem.buildPayload (/usr/local/lib/node_modules/decking/node_modules/dockerode/node_modules/docker-modem/lib/modem.js:134:15)
    at ClientRequest.<anonymous> (/usr/local/lib/node_modules/decking/node_modules/dockerode/node_modules/docker-modem/lib/modem.js:95:12)
    at ClientRequest.emit (events.js:117:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (http.js:1688:21)
    at HTTPParser.parserOnHeadersComplete [as onHeadersComplete] (http.js:121:23)
    at Socket.socketOnData [as ondata] (http.js:1583:20)
    at Pipe.onread (net.js:527:27)
```

Looking in the docker logs you see the following confirming this,

```
[d82a2482] +job build()
Dockerfile cannot be empty
[d82a2482] -job build() = ERR (1)
[error] server.go:1048 Error making handler: Dockerfile cannot be empty
[error] server.go:90 HTTP Error: statusCode=500 Dockerfile cannot be empty
```

The solution is to make sure `docker.buildImage()` is not called until the Dockerfile is copied and the context is full tarred up.

This is also seems to be the cause of #35.

--- a/src/decking.coffee
+++ b/src/decking.coffee
@@ -355,36 +355,40 @@ class Decking
 
     # @TODO for now, always assume we want to build from a Dockerfile
     # @TODO need a lot of careful validation here
-    fs.createReadStream(targetPath).pipe fs.createWriteStream("./Dockerfile")
+    readStream = fs.createReadStream(targetPath)
+    writeStream = fs.createWriteStream("./Dockerfile")
+    writeStream.on "done", () =>
 
-    options =
-      t: image
+      options =
+        t: image
 
-    if @hasArg "--no-cache"
-      log "Not using image cache"
-      options.nocache = true
+      if @hasArg "--no-cache"
+        log "Not using image cache"
+        options.nocache = true
 
-    log "Uploading compressed context..."
+      log "Uploading compressed context..."
 
-    # @TODO allow user to specifiy --exclude params to avoid unnecessarily huge tarballs
-    tar = child_process.spawn "tar", ["-c", "-", "./"]
+      # @TODO allow user to specifiy --exclude params to avoid unnecessarily huge tarballs
+      tar = child_process.spawn "tar", ["-c", "-", "./"]
 
-    docker.buildImage tar.stdout, options, (err, res) ->
-      fs.unlink "./Dockerfile", (err) -> log "[WARN] Could not remove Dockerfile" if err
+      docker.buildImage tar.stdout, options, (err, res) ->
+        fs.unlink "./Dockerfile", (err) -> log "[WARN] Could not remove Dockerfile" if err
 
-      return done err if err
+        return done err if err
 
-      if res.headers["content-type"] is "application/json"
-        res
-          .pipe(JSONStream.parse "stream")
-          .pipe(process.stdout)
-      else
-        # we don't need an if/else but let's keep it for clarity; it'd be too easy to
-        # skim-read the code and misinterpret the first pipe otherwise
-        res
-          .pipe(process.stdout)
+        if res.headers["content-type"] is "application/json"
+          res
+            .pipe(JSONStream.parse "stream")
+            .pipe(process.stdout)
+        else
+          # we don't need an if/else but let's keep it for clarity; it'd be too easy to
+          # skim-read the code and misinterpret the first pipe otherwise
+          res
+            .pipe(process.stdout)
 
-      res.on "end", done
+        res.on "end", done
+
+    readStream.pipe writeStream
 
   execute: (done) ->
     @command = "help" if not @command or @command is "-h" or @command is "--help"

